README.md
# Snowplow Snowflake Loader

## Overview

Snowflake Loader is a project to load Snowplow enriched events into the [Snowflake cloud data warehouse][snowflake]. It consists on two independent applications:

- *Snowplow Snowflake Transformer* - a Spark job responsible for transforming enriched events into a Snowflake-compatible format;
- *Snowplow Snowflake Loader* - a CLI application responsible for loading Snowplow-compatible enriched events into Snowflake DB.

| **[Technical Docs][techdocs]**    | **[Setup Guide][setup]**    | **[Contributing][contributing]**          |
|-----------------------------------|-----------------------------|-------------------------------------------|
| [![i1][techdocs-image]][techdocs] | [![i2][setup-image]][setup] | [![i3][contributing-image]][contributing] |

## Quickstart

Assuming git, **[Vagrant][vagrant-install]** and **[VirtualBox][virtualbox-install]** installed:

```bash
host$ git clone https://github.com/snowplow-product/snowplow-snowflake-loader.git
host$ cd snowplow-snowflake-loader
host$ vagrant up && vagrant ssh
guest$ cd /vagrant
guest$ sbt test
```

## Copyright and License

Snowflake Loader is copyright 2017-2018 Snowplow Analytics Ltd.

Licensed under the **[Apache License, Version 2.0][license]** (the "License");
you may not use this software except in compliance with the License.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[vagrant-install]: http://docs.vagrantup.com/v2/installation/index.html
[virtualbox-install]: https://www.virtualbox.org/wiki/Downloads

[license-image]: http://img.shields.io/badge/license-Apache--2-blue.svg?style=flat
[license]: http://www.apache.org/licenses/LICENSE-2.0

[travis]: https://travis-ci.org/snowplow/snowplowsnowflaketransformer
[travis-image]: https://travis-ci.org/snowplow/snowplowsnowflaketransformer.png?branch=master

[release-image]: http://img.shields.io/badge/release-0.4.0-blue.svg?style=flat
[releases]: https://github.com/snowplow/snowplowsnowflaketransformer/releases

[snowflake]: https://www.snowflake.com/

[techdocs]: https://github.com/snowplow-incubator/snowplow-snowflake-loader/wiki/
[techdocs-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/techdocs.png

[setup]: https://github.com/snowplow-incubator/snowplow-snowflake-loader/wiki/Setup-Guide
[setup-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/setup.png

[contributing]: https://github.com/snowplow-incubator/snowplow-snowflake-loader/wiki/
[contributing-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/contributing.png


CHANGELOG
Version 0.4.0 (2019-01-02)
--------------------------
Add in-batch natural deduplication (#48)
Bump Snowflake JDBC driver to 3.6.22 (#69)
Add table migration command (#68)
Recover DynamoDB on expired token exception (#51)
Widen geo_region to 3 characters (#58)
Fix silent skip for folder outside of a stage (#62)
Add a link to wiki to README (#65)
Remove VARCHAR-precision when loading from temporary table (#54)
Update JDBC useragent to self-identify Snowplow (#52)
Widen sensitive columns in atomic.events to support pseudonymization (#56)
Add optional host property to configuration (#59)
Truncate columns on loading (#55)
Update README markdown in according with CommonMark (#46)
Add ON_ERROR setting (#43)
Setup of warehouse should set Auto Suspend to 5 mins and Auto Resume to true (#15)
Use s3committer in Transformer (#67)
Add cross-batch natural deduplication (#33)

Version 0.3.1 (2018-02-02)
--------------------------
Remove part-* pattern from Transformer (#44)

Version 0.3.0 (2017-12-28)
--------------------------
Allow loader to use temporary credentials (#37)
Allow JDBC credentials to be stored in EC2 parameter store (#36)
Allow loader to use the full AWS credentials chain (#32)
Fix loading of NULLs as strings (#14)
Use self-describing configuration (#19)
Use transaction for altering table and load (#25)
Add AWS credentials to stage (#20)
Add --dry-run option (#9)
Remove all references to Project Strawberry (#16)
Change License to Apache 2.0 (#34)
Bump Snowflake JDBC driver to 3.3.2 (#29)
Bump Scala Analytics SDK to 0.2.1 (#27)
Bump AWS SDK to 1.11.208 (#26)
Make SparkContext initializing eagerly (#24)
Re-encrypt AWS credentials in .travis.yml (#40)

Version 0.2.0 (2017-10-05)
--------------------------
Change file format to JSON (#7)
Add preliminary checks for Snowflake entities (#11)
Add support for other regions (#6)
Separate AWS region from Snowflake region (#12)
Add deployment to all snowplow-hosted-assets buckets (#4)
Bump Snowflake JDBC to 3.2.5 (#2)
Improve load error message (#5)
Create test specification to check atomic.events DDL correctness (#8)
Refactor getUnprocessed to use DynamoDB scan (#17)

Version 0.1.0 (2017-08-31)
--------------------------
Initial release


loader/src/main/scala/com.snowplowanalytics.snowflake.loader/ast/AtomicDef.scala
/*
 * Copyright (c) 2017 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowflake.loader.ast

import com.snowplowanalytics.snowflake.loader.ast.CreateTable._
import com.snowplowanalytics.snowflake.loader.ast.SnowflakeDatatype._

object AtomicDef {
  /**
   * List of columns for enriched event
   */
  val columns = List(
    // App
    Column("app_id", Varchar(Some(255))),
    Column("platform", Varchar(Some(255))),

    // Data/time
    Column("etl_tstamp", Timestamp),
    Column("collector_tstamp", Timestamp, notNull = true),
    Column("dvce_created_tstamp", Timestamp),

    // Event
    Column("event", Varchar(Some(128))),
    Column("event_id", Char(36), notNull = true, unique = true),
    Column("txn_id", Integer),

    // Namespacing and versioning
    Column("name_tracker", Varchar(Some(128))),
    Column("v_tracker", Varchar(Some(100))),
    Column("v_collector", Varchar(Some(100)), notNull = true),
    Column("v_etl", Varchar(Some(100)), notNull = true),

    // User id and visit
    Column("user_id", Varchar(Some(255))),
    Column("user_ipaddress", Varchar(Some(128))),
    Column("user_fingerprint", Varchar(Some(128))),
    Column("domain_userid", Varchar(Some(128))),
    Column("domain_sessionidx", SmallInt),
    Column("network_userid", Varchar(Some(128))),

    // Location
    Column("geo_country", Char(2)),
    Column("geo_region", Char(2)),
    Column("geo_city", Varchar(Some(75))),
    Column("geo_zipcode", Varchar(Some(15))),
    Column("geo_latitude", DoublePrecision),
    Column("geo_longitude", DoublePrecision),
    Column("geo_region_name", Varchar(Some(100))),

    // Ip lookups
    Column("ip_isp", Varchar(Some(100))),
    Column("ip_organization", Varchar(Some(128))),
    Column("ip_domain", Varchar(Some(128))),
    Column("ip_netspeed", Varchar(Some(100))),

    // Page
    Column("page_url", Varchar(Some(4096))),
    Column("page_title", Varchar(Some(2000))),
    Column("page_referrer", Varchar(Some(4096))),

    // Page URL components
    Column("page_urlscheme", Varchar(Some(16))),
    Column("page_urlhost", Varchar(Some(255))),
    Column("page_urlport", Integer),
    Column("page_urlpath", Varchar(Some(3000))),
    Column("page_urlquery", Varchar(Some(6000))),
    Column("page_urlfragment", Varchar(Some(3000))),

    // Referrer URL components
    Column("refr_urlscheme", Varchar(Some(16))),
    Column("refr_urlhost", Varchar(Some(255))),
    Column("refr_urlport", Integer),
    Column("refr_urlpath", Varchar(Some(6000))),
    Column("refr_urlquery", Varchar(Some(6000))),
    Column("refr_urlfragment", Varchar(Some(3000))),

    // Referrer details
    Column("refr_medium", Varchar(Some(25))),
    Column("refr_source", Varchar(Some(50))),
    Column("refr_term", Varchar(Some(255))),

    // Marketing
    Column("mkt_medium", Varchar(Some(255))),
    Column("mkt_source", Varchar(Some(255))),
    Column("mkt_term", Varchar(Some(255))),
    Column("mkt_content", Varchar(Some(500))),
    Column("mkt_campaign", Varchar(Some(255))),

    // Custom structured event
    Column("se_category", Varchar(Some(1000))),
    Column("se_action", Varchar(Some(1000))),
    Column("se_label", Varchar(Some(1000))),
    Column("se_property", Varchar(Some(1000))),
    Column("se_value", DoublePrecision),

    // Ecommerce
    Column("tr_orderid", Varchar(Some(255))),
    Column("tr_affiliation", Varchar(Some(255))),
    Column("tr_total", Number(18,2)),
    Column("tr_tax", Number(18,2)),
    Column("tr_shipping", Number(18,2)),
    Column("tr_city", Varchar(Some(255))),
    Column("tr_state", Varchar(Some(255))),
    Column("tr_country", Varchar(Some(255))),
    Column("ti_orderid", Varchar(Some(255))),
    Column("ti_sku", Varchar(Some(255))),
    Column("ti_name", Varchar(Some(255))),
    Column("ti_category", Varchar(Some(255))),
    Column("ti_price", Number(18,2)),
    Column("ti_quantity", Integer),

    // Page ping
    Column("pp_xoffset_min", Integer),
    Column("pp_xoffset_max", Integer),
    Column("pp_yoffset_min", Integer),
    Column("pp_yoffset_max", Integer),

    // Useragent
    Column("useragent", Varchar(Some(1000))),

    // Browser
    Column("br_name", Varchar(Some(50))),
    Column("br_family", Varchar(Some(50))),
    Column("br_version", Varchar(Some(50))),
    Column("br_type", Varchar(Some(50))),
    Column("br_renderengine", Varchar(Some(50))),
    Column("br_lang", Varchar(Some(255))),
    Column("br_features_pdf", Boolean),
    Column("br_features_flash", Boolean),
    Column("br_features_java", Boolean),
    Column("br_features_director", Boolean),
    Column("br_features_quicktime", Boolean),
    Column("br_features_realplayer", Boolean),
    Column("br_features_windowsmedia", Boolean),
    Column("br_features_gears", Boolean),
    Column("br_features_silverlight", Boolean),
    Column("br_cookies", Boolean),
    Column("br_colordepth", Varchar(Some(12))),
    Column("br_viewwidth", Integer),
    Column("br_viewheight", Integer),

    // Operating System
    Column("os_name", Varchar(Some(50))),
    Column("os_family", Varchar(Some(50))),
    Column("os_manufacturer", Varchar(Some(50))),
    Column("os_timezone", Varchar(Some(255))),

    // Device/Hardware
    Column("dvce_type", Varchar(Some(50))),
    Column("dvce_ismobile", Boolean),
    Column("dvce_screenwidth", Integer),
    Column("dvce_screenheight", Integer),

    // Document
    Column("doc_charset", Varchar(Some(128))),
    Column("doc_width", Integer),
    Column("doc_height", Integer),

    // Currency
    Column("tr_currency", Char(3)),
    Column("tr_total_base", Number(18,2)),
    Column("tr_tax_base", Number(18,2)),
    Column("tr_shipping_base", Number(18,2)),
    Column("ti_currency", Char(3)),
    Column("ti_price_base", Number(18,2)),
    Column("base_currency", Char(3)),

    // Geolocation
    Column("geo_timezone", Varchar(Some(64))),

    // Click ID
    Column("mkt_clickid", Varchar(Some(128))),
    Column("mkt_network", Varchar(Some(64))),

    // ETL Tags
    Column("etl_tags", Varchar(Some(500))),

    // Time event was sent
    Column("dvce_sent_tstamp", Timestamp),

    // Referer
    Column("refr_domain_userid", Varchar(Some(128))),
    Column("refr_dvce_tstamp", Timestamp),

    // Session ID
    Column("domain_sessionid", Char(128)),

    // Derived timestamp
    Column("derived_tstamp", Timestamp),

    // Event schema
    Column("event_vendor", Varchar(Some(1000))),
    Column("event_name", Varchar(Some(1000))),
    Column("event_format", Varchar(Some(128))),
    Column("event_version", Varchar(Some(128))),

    // Event fingerprint
    Column("event_fingerprint", Varchar(Some(128))),

    // True timestamp
    Column("true_tstamp", Timestamp)
  )

  /** Get statement to create standard table with custom schema */
  def getTable(schema: String = Defaults.Schema): CreateTable =
    CreateTable(schema, Defaults.Table, columns, Some(PrimaryKeyConstraint("event_id_pk", "event_id")))
}
